From 5c45c0ba41955f7b70278be85fed9200484efd21 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D0=90=D0=BB=D0=B5=D0=BA=D1=81=D0=B5=D0=B9?=
 <alexey.pawlow@gmail.com>
Date: Thu, 17 Jun 2021 18:52:07 +0530
Subject: [PATCH 035/N] getcompiler: expose MINGW toolchain related
 information

* include MINGW and UCRT
* include things like "64 bit (AMD64)" for compat with the MSVC build
* We include "GCC" even for Clang builds since various code out there
  checks for GCC in the string to detect a MINGW build.
---
 Python/getcompiler.c | 44 ++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 42 insertions(+), 2 deletions(-)

diff --git a/Python/getcompiler.c b/Python/getcompiler.c
index a5d2623..5a9cf76 100644
--- a/Python/getcompiler.c
+++ b/Python/getcompiler.c
@@ -7,10 +7,50 @@
 
 // Note the __clang__ conditional has to come before the __GNUC__ one because
 // clang pretends to be GCC.
-#if defined(__clang__)
+#if defined(__clang__) && !defined(_WIN32)
 #define COMPILER "[Clang " __clang_version__ "]"
 #elif defined(__GNUC__)
-#define COMPILER "[GCC " __VERSION__ "]"
+/* To not break compatibility with things that determine
+   CPU arch by calling get_build_version in msvccompiler.py
+   (such as NumPy) add "32 bit" or "64 bit (AMD64)" on Windows
+   and also use a space as a separator rather than a newline. */
+#if defined(_WIN32)
+#  define COMP_SEP " "
+#  if defined(__x86_64__)
+#    define ARCH_SUFFIX " 64 bit (AMD64)"
+#  elif defined(__aarch64__)
+#    define ARCH_SUFFIX " 64 bit (ARM64)"
+#  elif defined(__arm__)
+#    define ARCH_SUFFIX " 32 bit (ARM)"
+#  else
+#    define ARCH_SUFFIX " 32 bit"
+#  endif
+#else
+#  define COMP_SEP "\n"
+#  define ARCH_SUFFIX ""
+#endif
+
+#if defined __MINGW32__
+#  define ARCH_PREFIX "MINGW "
+#else
+#  define ARCH_PREFIX ""
+#endif
+
+#if defined(__clang__)
+#  if defined(_UCRT)
+// Some code checks for "GCC" in the compiler string, so we include it here.
+#    define COMPILER COMP_SEP "[" ARCH_PREFIX "Clang UCRT " __clang_version__ ARCH_SUFFIX " (GCC)]"
+#  else
+#    define COMPILER COMP_SEP "[" ARCH_PREFIX "Clang " __clang_version__ ARCH_SUFFIX " (GCC)]"
+#  endif
+#else
+#  if defined(_UCRT)
+#    define COMPILER COMP_SEP "[" ARCH_PREFIX "GCC UCRT " __VERSION__ ARCH_SUFFIX "]"
+#  else
+#    define COMPILER COMP_SEP "[" ARCH_PREFIX "GCC " __VERSION__ ARCH_SUFFIX "]"
+#  endif
+#endif
+
 // Generic fallbacks.
 #elif defined(__cplusplus)
 #define COMPILER "[C++]"
